src/TMap.cpp

```diff
@@ -2649,6 +2653,542 @@ void TMap::setRoomNamesShown(bool shown)
     setUserDataBool(mUserData, ROOM_UI_SHOWNAME, shown);
 }
 
+std::pair<bool, QString> TMap::writeJsonMapFile(const QString& dest)
+{
+    QString destination{dest};
+    if (!dest.endsWith(QLatin1String(".json"), Qt::CaseInsensitive)) {
+        destination.append(QLatin1String(".json"));
+    }
+
+    if (mpProgressDialog) {
+        return {false, QStringLiteral("import or export already in progress")};
+    }
+
+    mProgressDialogRoomsTotal = mpRoomDB->getRoomMap().count();
+    mProgressDialogAreasTotal = mpRoomDB->getAreaMap().count();
+    QMapIterator<int, TArea*> itArea(mpRoomDB->getAreaMap());
+    mProgressDialogLabelsTotal = 0;
+    while (itArea.hasNext()) {
+        itArea.next();
+        auto pA = itArea.value();
+        if (pA) {
+            mProgressDialogLabelsTotal += pA->mMapLabels.size();
+        }
+    }
+
+    mpProgressDialog = new QProgressDialog(tr("Exporting JSON map data from %1\n"
+                                              "Areas: %2 of: %3   Rooms: %4 of: %5   Labels: %6 of: %7...")
+                                           .arg(mProfileName,
+                                                QLatin1String("0"),
+                                                QString::number(mProgressDialogAreasTotal),
+                                                QLatin1String("0"),
+                                                QString::number(mProgressDialogRoomsTotal),
+                                                QLatin1String("0"),
+                                                QString::number(mProgressDialogLabelsTotal)),
+                                           tr("Abort"),
+                                           0,
+                                           mProgressDialogRoomsTotal,
+                                           mpHost->mpConsole);
+    mpProgressDialog->setValue(0);
+    mpProgressDialog->setWindowModality(Qt::NonModal);
+    mpProgressDialog->setWindowTitle(tr("Map JSON export", "This is a title of a progress window."));
+    mpProgressDialog->setWindowIcon(QIcon(QStringLiteral(":/icons/mudlet_map_download.png")));
+    mpProgressDialog->setMinimumWidth(500);
+    mpProgressDialog->setAutoClose(false);
+    mpProgressDialog->setAutoReset(false);
+    mpProgressDialog->setMinimumDuration(0); // Normally waits for 4 seconds before showing
+    qApp->processEvents();
+    QFile file(destination);
+    if (!file.open(QFile::OpenMode(QFile::Text|QFile::WriteOnly))) {
+        qWarning().noquote().nospace() << "TMap::writeJsonMapFile(...) WARNING - Could not open save file \"" << destination << "\".";
+        mpProgressDialog->setAttribute(Qt::WA_DeleteOnClose, true);
```

Comment:
Shouldn't this be set a creation?

Reply:
:thinking: TBH I am not sure - I can't recall whether the dialogue is decorated with a title bar with a close button, need to see if it has and then how things behave if I close it prematurely...

Reply:
It *seems* that the close button on the titlebar is the same as the "Cancel" button...